src/idpf.rs

```diff
@@ -0,0 +1,1831 @@
+//! This module implements the incremental distributed point function (IDPF) described in
+//! [[draft-irtf-cfrg-vdaf-03]].
+//!
+//! [draft-irtf-cfrg-vdaf-03]: https://datatracker.ietf.org/doc/draft-irtf-cfrg-vdaf/03/
+
+use crate::{
+    codec::{CodecError, Decode, Encode, ParameterizedDecode},
+    field::{Field255, Field64, FieldElement},
+    prng::Prng,
+    vdaf::{
+        prg::{Prg, RandSource, Seed, SeedStream},
+        VdafError, VERSION,
+    },
+};
+use bitvec::{bitvec, boxed::BitBox, prelude::Lsb0, slice::BitSlice, vec::BitVec, view::BitView};
+use std::{
+    collections::{HashMap, VecDeque},
+    fmt::Debug,
+    io::{Cursor, Read},
+    ops::Index,
+};
+use subtle::{Choice, ConditionallySelectable, ConstantTimeEq};
+
+/// IDPF-related errors.
+#[derive(Debug, thiserror::Error)]
+pub enum IdpfError {
+    /// Error from incompatible shares at different levels.
+    #[error("tried to merge shares from incompatible levels")]
+    MismatchedLevel,
+
+    /// Invalid parameter, indicates an invalid input to either [`gen`] or [`eval`].
+    #[error("invalid parameter: {0}")]
+    InvalidParameter(String),
+}
+
+/// An index used as the input to an IDPF evaluation.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct IdpfInput {
+    /// The index as a boxed bit slice.
+    index: BitBox,
+}
+
+impl IdpfInput {
+    /// Convert a slice of bytes into an IDPF input, where the bits of each byte are processed in
+    /// LSB-to-MSB order. (Subsequent bytes are processed in their natural order.)
+    pub fn from_bytes(bytes: &[u8]) -> IdpfInput {
+        let bit_slice_u8_storage = bytes.view_bits::<Lsb0>();
+        let mut bit_vec_usize_storage = bitvec![0; bit_slice_u8_storage.len()];
+        bit_vec_usize_storage.clone_from_bitslice(bit_slice_u8_storage);
+        IdpfInput {
+            index: bit_vec_usize_storage.into_boxed_bitslice(),
+        }
+    }
+
+    /// Convert a slice of booleans into an IDPF input.
+    pub fn from_bools(bools: &[bool]) -> IdpfInput {
```

Comment:
Seems odd that we have `From<BitVec>` and `From<BitBox>` below but these aren't `From<&[u8]>` and `From<&[bool]>`, but I'm OK with either.

Reply:
This was done to avoid implicit copies in from/into conversions, see [this thread](https://github.com/divviup/libprio-rs/pull/414#discussion_r1072891268).